File: exercises/python/11_three_address_code/three_address_code.py
"""
================================================================================
Exercise 11: Three-Address Code (TAC) Generation
================================================================================

CONCEPT:
    Three-Address Code is an intermediate representation used in compilers where
    each instruction has at most three operands (two sources and one destination).
    Complex expressions are broken down into a sequence of simple instructions,
    each involving at most one operator. Temporary variables (t1, t2, ...) hold
    intermediate results.

ALGORITHM:
    This implementation uses a recursive descent parser that respects operator
    precedence. The grammar is:
        assignment  -> IDENTIFIER '=' expression
        expression  -> term (('+' | '-') term)*
        term        -> factor (('*' | '/') factor)*
        factor      -> IDENTIFIER | NUMBER | '(' expression ')'

    As the parser recognizes each binary operation, it emits a three-address
    code instruction using a fresh temporary variable. The parser returns the
    name of the temporary (or the leaf operand) so that higher-level rules can
    reference the result. This naturally produces code in correct evaluation
    order with proper precedence.

TIME COMPLEXITY:  O(n) where n is the length of the expression (single pass).
SPACE COMPLEXITY: O(n) for the generated TAC instructions and temporaries.

EXAMPLE:
    Input:  x = a + b * c - d / e
    Output:
        t1 = b * c
        t2 = a + t1
        t3 = d / e
        t4 = t2 - t3
        x = t4
================================================================================
"""


class Lexer:
    """Tokenizes an expression string into a stream of tokens."""

    def __init__(self, text):
        self.text = text
        self.pos = 0
        self.tokens = []
        self._tokenize()
        self.current = 0

    def _tokenize(self):
        """Scan the input and produce a list of (type, value) tokens."""
        i = 0
        while i < len(self.text):
            ch = self.text[i]
            if ch.isspace():
                i += 1
            elif ch in "+-*/=()":
                self.tokens.append(("OP", ch))
                i += 1
            elif ch.isalpha() or ch == "_":
                start = i
                while i < len(self.text) and (self.text[i].isalnum() or self.text[i] == "_"):
                    i += 1
                self.tokens.append(("ID", self.text[start:i]))
            elif ch.isdigit():
                start = i
                while i < len(self.text) and self.text[i].isdigit():
                    i += 1
                self.tokens.append(("NUM", self.text[start:i]))
            else:
                raise SyntaxError(f"Unexpected character: '{ch}' at position {i}")
        self.tokens.append(("EOF", None))

    def peek(self):
        """Return the current token without consuming it."""
        return self.tokens[self.current]

    def advance(self):
        """Consume and return the current token."""
        token = self.tokens[self.current]
        self.current += 1
        return token

    def expect(self, token_type, value=None):
        """Consume the current token, raising an error if it does not match."""
        token = self.advance()
        if token[0] != token_type or (value is not None and token[1] != value):
            raise SyntaxError(
                f"Expected ({token_type}, {value}), got ({token[0]}, {token[1]})"
            )
        return token


class TACGenerator:
    """
    Recursive descent parser that emits three-address code while parsing.
    Operator precedence is encoded in the grammar:
        expression handles + and -
        term       handles * and /
        factor     handles identifiers, numbers, and parenthesised sub-expressions
    """

    def __init__(self):
        self.temp_counter = 0
        self.instructions = []

    def new_temp(self):
        """Allocate a fresh temporary variable name."""
        self.temp_counter += 1
        return f"t{self.temp_counter}"

    def emit(self, result, left, operator, right):
        """Append a three-address code instruction."""
        self.instructions.append((result, left, operator, right))

    # -------------------------------------------------------------------------
    # Parsing rules
    # -------------------------------------------------------------------------

    def parse_assignment(self, lexer):
        """assignment -> ID '=' expression"""
        name_token = lexer.expect("ID")
        target = name_token[1]
        lexer.expect("OP", "=")
        expr_result = self.parse_expression(lexer)

        # Emit the final assignment
        self.instructions.append((target, expr_result, None, None))
        return target

    def parse_expression(self, lexer):
        """expression -> term (('+' | '-') term)*"""
        left = self.parse_term(lexer)

        while lexer.peek()[0] == "OP" and lexer.peek()[1] in ("+", "-"):
            operator = lexer.advance()[1]
            right = self.parse_term(lexer)
            temp = self.new_temp()
            self.emit(temp, left, operator, right)
            left = temp

        return left

    def parse_term(self, lexer):
        """term -> factor (('*' | '/') factor)*"""
        left = self.parse_factor(lexer)

        while lexer.peek()[0] == "OP" and lexer.peek()[1] in ("*", "/"):
            operator = lexer.advance()[1]
            right = self.parse_factor(lexer)
            temp = self.new_temp()
            self.emit(temp, left, operator, right)
            left = temp

        return left

    def parse_factor(self, lexer):
        """factor -> ID | NUM | '(' expression ')'"""
        token = lexer.peek()

        if token[0] == "ID":
            lexer.advance()
            return token[1]
        elif token[0] == "NUM":
            lexer.advance()
            return token[1]
        elif token[0] == "OP" and token[1] == "(":
            lexer.advance()  # consume '('
            result = self.parse_expression(lexer)
            lexer.expect("OP", ")")
            return result
        else:
            raise SyntaxError(f"Unexpected token: {token}")

    # -------------------------------------------------------------------------
    # Public interface
    # -------------------------------------------------------------------------

    def generate(self, expression):
        """Parse an expression string and return three-address code."""
        self.instructions = []
        self.temp_counter = 0
        lexer = Lexer(expression)
        self.parse_assignment(lexer)
        return self.instructions

    def format_instructions(self):
        """Return a human-readable string of all TAC instructions."""
        lines = []
        for instr in self.instructions:
            result, left, operator, right = instr
            if operator is None:
                # Simple copy / assignment
                lines.append(f"    {result} = {left}")
            else:
                lines.append(f"    {result} = {left} {operator} {right}")
        return "\n".join(lines)


def process_expression(generator, expression):
    """Generate and display TAC for a single expression."""
    print(f"  Expression : {expression}")
    print(f"  Three-Address Code:")
    generator.generate(expression)
    print(generator.format_instructions())
    print()


def main():
    print("=" * 65)
    print("   THREE-ADDRESS CODE GENERATION")
    print("=" * 65)
    print()

    generator = TACGenerator()

    test_expressions = [
        "x = a + b * c - d / e",
        "y = (a + b) * (c - d)",
        "z = a * b + c * d + e * f",
        "w = a + b",
        "r = (a + b) * c - d / (e + f)",
    ]

    for i, expr in enumerate(test_expressions, start=1):
        print(f"  --- Test Case {i} ---")
        process_expression(generator, expr)

    print("=" * 65)
    print("  All expressions processed successfully.")
    print("=" * 65)


if __name__ == "__main__":
    main()

File: exercises/c/11_three_address_code/three_address_code.c
/*
 * ===========================================================================
 * Exercise 11: Three-Address Code (TAC) Generation  --  C Implementation
 * ===========================================================================
 *
 * CONCEPT:
 *   Three-Address Code is an intermediate representation used in compilers
 *   where each instruction has at most three operands (two sources and one
 *   destination).  Complex expressions are decomposed into a sequence of
 *   simple instructions, each involving at most one arithmetic operator.
 *   Temporary variables (t1, t2, ...) hold intermediate results.
 *
 * ALGORITHM:
 *   A recursive descent parser walks the token stream.  Operator precedence
 *   is encoded directly in the grammar:
 *       assignment  -> ID '=' expression
 *       expression  -> term (('+' | '-') term)*
 *       term        -> factor (('*' | '/') factor)*
 *       factor      -> ID | NUMBER | '(' expression ')'
 *   Each binary operation emits a TAC instruction with a fresh temporary
 *   and returns the temporary name so the caller can reference it.
 *
 * TIME COMPLEXITY:  O(n), single pass over the expression.
 * SPACE COMPLEXITY: O(n) for tokens and generated instructions.
 *
 * EXAMPLE:
 *   Input:  x = a + b * c - d / e
 *   Output:
 *       t1 = b * c
 *       t2 = a + t1
 *       t3 = d / e
 *       t4 = t2 - t3
 *       x  = t4
 * ===========================================================================
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

/* ---------- constants ----------------------------------------------------- */
#define MAX_TOKEN_LEN  64
#define MAX_TOKENS     256
#define MAX_TAC        256

/* ---------- token types --------------------------------------------------- */
typedef enum { TOK_ID, TOK_NUM, TOK_OP, TOK_EOF } TokenType;

typedef struct {
    TokenType type;
    char      value[MAX_TOKEN_LEN];
} Token;

/* ---------- TAC instruction ----------------------------------------------- */
typedef struct {
    char result[MAX_TOKEN_LEN];
    char left[MAX_TOKEN_LEN];
    char op;           /* '+', '-', '*', '/', or '\0' for simple copy */
    char right[MAX_TOKEN_LEN];
} TACInstr;

/* ---------- global state -------------------------------------------------- */
static Token    tokens[MAX_TOKENS];
static int      token_count = 0;
static int      token_pos   = 0;

static TACInstr tac[MAX_TAC];
static int      tac_count   = 0;
static int      temp_counter = 0;

/* ---------- utility functions --------------------------------------------- */

/* Create a new temporary name and write it into `buf`. */
static void new_temp(char *buf, int buf_size)
{
    temp_counter++;
    snprintf(buf, buf_size, "t%d", temp_counter);
}

/* Emit a TAC instruction. `op` may be '\0' for a simple copy. */
static void emit(const char *result, const char *left, char op, const char *right)
{
    if (tac_count >= MAX_TAC) {
        fprintf(stderr, "Error: TAC instruction limit exceeded.\n");
        exit(1);
    }
    TACInstr *instr = &tac[tac_count++];
    strncpy(instr->result, result, MAX_TOKEN_LEN - 1);
    strncpy(instr->left,   left,   MAX_TOKEN_LEN - 1);
    instr->op = op;
    if (right)
        strncpy(instr->right, right, MAX_TOKEN_LEN - 1);
    else
        instr->right[0] = '\0';
}

/* ---------- lexer --------------------------------------------------------- */

static void tokenize(const char *text)
{
    token_count = 0;
    int i = 0, len = (int)strlen(text);

    while (i < len) {
        if (isspace((unsigned char)text[i])) {
            i++;
        } else if (strchr("+-*/=()", text[i])) {
            tokens[token_count].type = TOK_OP;
            tokens[token_count].value[0] = text[i];
            tokens[token_count].value[1] = '\0';
            token_count++;
            i++;
        } else if (isalpha((unsigned char)text[i]) || text[i] == '_') {
            int start = i;
            while (i < len && (isalnum((unsigned char)text[i]) || text[i] == '_'))
                i++;
            int span = i - start;
            if (span >= MAX_TOKEN_LEN) span = MAX_TOKEN_LEN - 1;
            strncpy(tokens[token_count].value, text + start, span);
            tokens[token_count].value[span] = '\0';
            tokens[token_count].type = TOK_ID;
            token_count++;
        } else if (isdigit((unsigned char)text[i])) {
            int start = i;
            while (i < len && isdigit((unsigned char)text[i]))
                i++;
            int span = i - start;
            if (span >= MAX_TOKEN_LEN) span = MAX_TOKEN_LEN - 1;
            strncpy(tokens[token_count].value, text + start, span);
            tokens[token_count].value[span] = '\0';
            tokens[token_count].type = TOK_NUM;
            token_count++;
        } else {
            fprintf(stderr, "Error: unexpected character '%c'\n", text[i]);
            exit(1);
        }
    }
    tokens[token_count].type = TOK_EOF;
    tokens[token_count].value[0] = '\0';
    token_count++;
}

static Token *peek(void)
{
    return &tokens[token_pos];
}

static Token *advance(void)
{
    return &tokens[token_pos++];
}

static Token *expect(TokenType type, const char *value)
{
    Token *tok = advance();
    if (tok->type != type || (value && strcmp(tok->value, value) != 0)) {
        fprintf(stderr, "Error: expected (%d, %s), got (%d, %s)\n",
                type, value ? value : "?", tok->type, tok->value);
        exit(1);
    }
    return tok;
}

/* ---------- recursive descent parser -------------------------------------- */

/* Forward declarations. */
static void parse_expression(char *result_name);
static void parse_term(char *result_name);
static void parse_factor(char *result_name);

/*
 * factor -> ID | NUM | '(' expression ')'
 * Writes the operand name (variable, number, or temp) into result_name.
 */
static void parse_factor(char *result_name)
{
    Token *tok = peek();

    if (tok->type == TOK_ID) {
        advance();
        strcpy(result_name, tok->value);
    } else if (tok->type == TOK_NUM) {
        advance();
        strcpy(result_name, tok->value);
    } else if (tok->type == TOK_OP && tok->value[0] == '(') {
        advance(); /* consume '(' */
        parse_expression(result_name);
        expect(TOK_OP, ")");
    } else {
        fprintf(stderr, "Error: unexpected token '%s'\n", tok->value);
        exit(1);
    }
}

/*
 * term -> factor (('*' | '/') factor)*
 */
static void parse_term(char *result_name)
{
    char left[MAX_TOKEN_LEN];
    parse_factor(left);

    while (peek()->type == TOK_OP &&
           (peek()->value[0] == '*' || peek()->value[0] == '/')) {
        char op = advance()->value[0];
        char right[MAX_TOKEN_LEN];
        parse_factor(right);

        char temp[MAX_TOKEN_LEN];
        new_temp(temp, MAX_TOKEN_LEN);
        emit(temp, left, op, right);
        strcpy(left, temp);
    }

    strcpy(result_name, left);
}

/*
 * expression -> term (('+' | '-') term)*
 */
static void parse_expression(char *result_name)
{
    char left[MAX_TOKEN_LEN];
    parse_term(left);

    while (peek()->type == TOK_OP &&
           (peek()->value[0] == '+' || peek()->value[0] == '-')) {
        char op = advance()->value[0];
        char right[MAX_TOKEN_LEN];
        parse_term(right);

        char temp[MAX_TOKEN_LEN];
        new_temp(temp, MAX_TOKEN_LEN);
        emit(temp, left, op, right);
        strcpy(left, temp);
    }

    strcpy(result_name, left);
}

/*
 * assignment -> ID '=' expression
 */
static void parse_assignment(void)
{
    Token *name_tok = expect(TOK_ID, NULL);
    expect(TOK_OP, "=");

    char expr_result[MAX_TOKEN_LEN];
    parse_expression(expr_result);

    /* Final assignment */
    emit(name_tok->value, expr_result, '\0', NULL);
}

/* ---------- driver -------------------------------------------------------- */

static void generate_tac(const char *expression)
{
    tac_count    = 0;
    temp_counter = 0;
    token_pos    = 0;

    tokenize(expression);
    parse_assignment();
}

static void print_tac(void)
{
    for (int i = 0; i < tac_count; i++) {
        TACInstr *instr = &tac[i];
        if (instr->op != '\0')
            printf("    %s = %s %c %s\n",
                   instr->result, instr->left, instr->op, instr->right);
        else
            printf("    %s = %s\n", instr->result, instr->left);
    }
}

int main(void)
{
    const char *test_expressions[] = {
        "x = a + b * c - d / e",
        "y = (a + b) * (c - d)",
        "z = a * b + c * d + e * f",
        "w = a + b",
        "r = (a + b) * c - d / (e + f)"
    };
    int num_tests = sizeof(test_expressions) / sizeof(test_expressions[0]);

    printf("=================================================================\n");
    printf("   THREE-ADDRESS CODE GENERATION  (C)\n");
    printf("=================================================================\n\n");

    for (int i = 0; i < num_tests; i++) {
        printf("  --- Test Case %d ---\n", i + 1);
        printf("  Expression : %s\n", test_expressions[i]);
        printf("  Three-Address Code:\n");
        generate_tac(test_expressions[i]);
        print_tac();
        printf("\n");
    }

    printf("=================================================================\n");
    printf("  All expressions processed successfully.\n");
    printf("=================================================================\n");
    return 0;
}
